### PR TITLE
feat(expr): support date, interval and timestamp

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Fixes https://github.com/datafuselabs/databend/issues/6635
Blocked by https://github.com/datafuselabs/databend/pull/6867

### CODE REVIEW

File: common/expression/src/types/datetime.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+use crate::ScalarRef;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DateType;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IntervalType;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TimestampType;
```
Comment (sundy-li) [inline common/expression/src/types/datetime.rs:34]: we have a precision argument in `TimestampType`

File: common/expression/src/types/datetime.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    fn upcast_domain(domain: Self::Domain) -> Domain {
+        Domain::Int(domain)
+    }
+
+    fn full_domain() -> Self::Domain {
+        IntDomain {
+            min: i32::MIN as i64,
+            max: i32::MAX as i64,
```
Comment (PsiACE) [inline common/expression/src/types/datetime.rs:71]:
```
/// date ranges from 1000-01-01 to 9999-12-31
/// date_max and date_min means days offset from 1970-01-01
/// any date not in the range will be invalid
pub const DATE_MAX: i32 = 2932896;
pub const DATE_MIN: i32 = -354285;
```

File: common/expression/src/types/datetime.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    fn upcast_domain(domain: Self::Domain) -> Domain {
+        Domain::Int(domain)
+    }
+
+    fn full_domain() -> Self::Domain {
+        IntDomain {
+            min: i64::MIN,
+            max: i64::MAX,
```
Comment (PsiACE) [inline common/expression/src/types/datetime.rs:151]:
```
/// timestamp ranges from 1000-01-01 00:00:00.000000 to 9999-12-31 23:59:59.999999
/// timestamp_max and timestamp_min means days offset from 1970-01-01 00:00:00.000000
/// any timestamp not in the range will be invalid
pub const TIMESTAMP_MAX: i64 = 253402300799999999;
pub const TIMESTAMP_MIN: i64 = -30610224000000000;
```

**Comment by PsiACE** [comment]
also, may require some

```
Date => ArrowDataType::Date32,
Datetime(unit, tz) => ArrowDataType::Timestamp(unit.to_arrow(), tz.clone()),
```

**Comment by andylokandy** [comment]
@PsiACE Have you been working on it? If so I'd like to close this one.

**Comment by PsiACE** [comment]
> @PsiACE Have you been working on it? If so I'd like to close this one.

You can keep working on this, I haven't made that much progress, I just happened to notice these.

**Comment by drmingdrmer** [comment]
- `main` branch is force pushed, update your working branch please: #7261

**Comment by drmingdrmer** [comment]
`main` branch is force pushed(**again** `:(`, to cleanup big object in history), update your working branch please: [To everybody: main branch is force pushed. #7263](https://github.com/datafuselabs/databend/issues/7263)

**Comment by andylokandy** [comment]
Closing in favor of https://github.com/datafuselabs/databend/pull/7393

**Pull request closed without merging**